### PR TITLE
Adding retry mechanism for when command channel fails with ZMQErrors

### LINKED ISSUE #1146: "ZMQError: Operation cannot be accomplished in current state" triggers shutdown
## Brief description and request
I've been seeing an error crop up in longer running Parsl DFK's which trigger a graceful, but unwanted shutdown. I can't find a reliable/minimal example to cause this trigger other than "run a while." So, I'm going to post as much debug info I can about our setup and am seeking advice on things to test or other information which would be helpful in debugging this error.

I'm happy to test whatever is needed, I can also make direct code modifications 

## Backtrace of error thrown
 The final error is in the title but here is the backtrace, it reliably fails on this line:

```
... (there are higher functions in this stack from our program)
  File "/home/lnaden/github/qcfractal/qcfractal/queue/parsl_adapter.py", line 94, in count_running_workers
    if hasattr(executor, 'connected_workers'):
  File "/home/lnaden/miniconda3/envs/qca/lib/python3.7/site-packages/parsl/executors/high_throughput/executor.py", line 463, in connected_workers
    workers = self.command_client.run("WORKERS")
  File "/home/lnaden/miniconda3/envs/qca/lib/python3.7/site-packages/parsl/executors/high_throughput/zmq_pipes.py", line 40, in run
    self.zmq_socket.send_pyobj(message, copy=True)
  File "/home/lnaden/miniconda3/envs/qca/lib/python3.7/site-packages/zmq/sugar/socket.py", line 606, in send_pyobj
    return self.send(msg, flags=flags, **kwargs)
  File "/home/lnaden/miniconda3/envs/qca/lib/python3.7/site-packages/zmq/sugar/socket.py", line 395, in send
    return super(Socket, self).send(data, flags=flags, copy=copy, track=track)
  File "zmq/backend/cython/socket.pyx", line 725, in zmq.backend.cython.socket.Socket.send
  File "zmq/backend/cython/socket.pyx", line 772, in zmq.backend.cython.socket.Socket.send
  File "zmq/backend/cython/socket.pyx", line 247, in zmq.backend.cython.socket._send_copy
  File "zmq/backend/cython/socket.pyx", line 242, in zmq.backend.cython.socket._send_copy
  File "zmq/backend/cython/checkrc.pxd", line 25, in zmq.backend.cython.checkrc._check_rc
zmq.error.ZMQError: Operation cannot be accomplished in current state
```

Below this it behaves as though a standard shutdown signal and the next log item is `[I 190716 15:38:08 dflow:821] DFK cleanup initiated`

## Additional info

`if hasattr(executor, 'connected_workers'):` is inside a loop over the values: `parsl.dataflow.dflow.DataFlowKernel.executors.values()` 

Near as I can tell, this error is only thrown and logged at the higher level programs which are running Parsl and this does not appear in any of the logs or files from the `runinfo` folder. 

Leading up to the crash, there are multiple "loss of Manager" errors. The first set of lost managers happened ~6 hours after the processes were started, which is expected given the wall time of the cluster would have stopped them. However, after that initial set, we started loosing managers every 5-30 minutes seemingly randomly, all due to "too many missed heartbeats." This behavior is the same on 2 separate clusters (similar hardware, same SLURM scheduler). Not sure if this is related, but it's an observation.

Config file:
```python
Config(
    app_cache=True,
    checkpoint_files=None,
    checkpoint_mode=None,
    checkpoint_period=None,
    data_management_max_threads=10,
    executors=[HighThroughputExecutor(
        address='calogin1',
        cores_per_worker=16,
        heartbeat_period=30,
        heartbeat_threshold=120,
        interchange_port_range=(55000, 56000),
        label='QCFractal_Parsl_Slurm_Executor',
        launch_cmd='process_worker_pool.py {debug} {max_workers} -p {prefetch_capacity} -c {cores_per_worker} -m {mem_per_worker} --poll {poll_period} --task_url={task_url} --result_url={result_url} --logdir={logdir} --block_id={{block_id}} --hb_period={heartbeat_period} --hb_threshold={heartbeat_threshold} ',
        managed=True,
        max_workers=48,
        mem_per_worker=None,
        poll_period=10,
        prefetch_capacity=0,
        provider=SlurmProvider(
            'normal_q',
            channel=LocalChannel(envs={}, script_dir=None, userhome='/home/lnaden/qcarchive/run_cc'),
            cmd_timeout=10,
            exclusive=True,
            init_blocks=0,
            launcher=SingleNodeLauncher(),
            max_blocks=24,
            min_blocks=0,
            move_files=True,
            nodes_per_block=1,
            parallelism=1,
            scheduler_options='#SBATCH --exclusive\n#SBATCH -A themolssi\n',
            walltime='06:00:00',
            worker_init='source /home/lnaden/qcarchive/boot_qcarchive_env.sh'
        ),
        storage_access=[],
        suppress_failure=False,
        worker_debug=False,
        worker_logdir_root=None,
        worker_port_range=(54000, 55000),
        worker_ports=None,
        working_dir=None
    )],
    lazy_errors=True,
    monitoring=None,
    retries=0,
    run_dir='runinfo',
    strategy='simple',
    usage_tracking=False
)

```

Setup info: SLURM Cluster, executing locally on the head node in a background process. Directly calling and manipulating the `DataFlowKernel` after providing it a `Config`

### PR DESCRIPTION
**Pull request by yadudoc**
Reproducing #1146 is hard. Assuming that the ZMQError is coming out of a REQ/REP pipe deadlocking, destroying and recreating the context should be a reasonable fix.

Once this is reviewed, @Lnaden and @dgasmith, it would be very useful to get this tested over on your infrastructure.

Fixes #1146

### CODE REVIEW

File: parsl/executors/high_throughput/zmq_pipes.py [hunk 1]
```diff
@@ -37,8 +55,21 @@ def run(self, message):
...
+        try:
+            self.zmq_socket.send_pyobj(message, copy=True)
+            reply = self.zmq_socket.recv_pyobj()
+        except zmq.ZMQError:
+            logger.error("Potential ZMQ REQ-REP deadlock caught")
+            logger.info("Trying to reestablish context")
+            self.context.destroy()
+            self.context = zmq.Context()
+            self.port = self.create_socket_and_bind()
+            if max_retries > 0:
```
Comment (benclifford) [inline parsl/executors/high_throughput/zmq_pipes.py]: should this retries test be here? or at the start of run() or before line 66 which attempts a reconnect?

File: parsl/executors/high_throughput/zmq_pipes.py [hunk 2]
```diff
@@ -37,8 +55,21 @@ def run(self, message):
...
+                return self.run(message, max_retries=(max_retries - 1))
+            else:
+                logger.error("Command channel run reties exhausted. Unable to run command")
```
Comment (benclifford) [inline parsl/executors/high_throughput/zmq_pipes.py]: retRies

File: parsl/executors/high_throughput/zmq_pipes.py [hunk 3]
```diff
@@ -37,8 +55,21 @@ def run(self, message):
         in ZMQ sockets reaching a broken state once there are ~10k tasks in flight.
         This issue can be magnified if each the serialized buffer itself is larger.
         """
-        self.zmq_socket.send_pyobj(message, copy=True)
-        reply = self.zmq_socket.recv_pyobj()
+        try:
+            self.zmq_socket.send_pyobj(message, copy=True)
+            reply = self.zmq_socket.recv_pyobj()
+        except zmq.ZMQError:
+            logger.error("Potential ZMQ REQ-REP deadlock caught")
```
Comment (benclifford) [inline parsl/executors/high_throughput/zmq_pipes.py]: is there any more interesting info stored in the ZMQError (now or in future) that would be useful for debugging this? if so, log at least the message from it?

File: parsl/executors/high_throughput/zmq_pipes.py [hunk 4]
```diff
@@ -23,12 +23,30 @@ def __init__(self, ip_address, port_range):
...
         """
         self.context = zmq.Context()
-        self.zmq_socket = self.context.socket(zmq.REQ)
-        self.port = self.zmq_socket.bind_to_random_port("tcp://{}".format(ip_address),
-                                                        min_port=port_range[0],
-                                                        max_port=port_range[1])
+        self.ip_address = ip_address
+        self.port_range = port_range
+        self.port = None
+        self.port = self.create_socket_and_bind()
```
Comment (benclifford) [inline parsl/executors/high_throughput/zmq_pipes.py]: this looks weird - make create_socket_and_bind change self.port itself. maybe also make it create zmq.Context inside there.

**Comment by dgasmith** [comment]
Happy to try this out. This can either happen after an hour or a few days, so it might be a bit after we spin up runs to see if this fixes the issue.

File: parsl/executors/high_throughput/zmq_pipes.py (hunk 3 above)
Comment (yadudoc) [reply parsl/executors/high_throughput/zmq_pipes.py]: Should've used logger.exception instead that would capture this info. Will switch.

**Comment by mattwelborn** [comment]
@yadudoc @benclifford Thank you for your efforts on this. Let me know when it's ready to go, and I will test it on the QCArchive stack to see if #1146 continues to occur.

File: parsl/executors/high_throughput/zmq_pipes.py (hunk 4 above)
Comment (dgasmith) [reply parsl/executors/high_throughput/zmq_pipes.py]: This also resets port to ‘None’ at the moment as create doesn’t have a return value.

Comment (yadudoc) [reply parsl/executors/high_throughput/zmq_pipes.py]: Fixed.

**Comment by mattwelborn** [comment]
I tried this PR out with the latest QCFractal stack. When my tasks finish, I get `parsl.executors.high_throughput.interchange.BadRegistration: Manager b'63cc5b93b74d' attempted to register with a bad registration message. Caused a critical failure`. The full log (with a traceback at the end) is attached.

I also went back and tried the latest QCFractal with Parsl 0.8.0. That did not show the BadRegistration error, but did reproduce issue #1146.

[dt_parsl1321_1.txt](https://github.com/Parsl/parsl/files/3684119/dt_parsl1321_1.txt)

**Comment by mattwelborn** [comment]
Runinfo attached.
[runinfo.tar.gz](https://github.com/Parsl/parsl/files/3684124/runinfo.tar.gz)

**Comment by yadudoc** [comment]
@mattwelborn BadRegistration is a separate issue. Could you retry with `HighThroughputExecutor(suppress_failure=True)`.

**Comment by mattwelborn** [comment]
@yadudoc Thanks for the speedy reply. As instructed, I retried with `suppress_failure=True`. The system gets stuck in a state where a slurm job is running, but the tasks do not get run. The relevant output is:

`[D 191002 22:04:59 strategy:200] Executor QCFractal_Parsl_Slurm_Executor has 16 active tasks, 1/0/0 running/submitted/pending blocks, and 0 connected workers`

**Comment by yadudoc** [comment]
Could you share the tarball of the rundir for this failure ?

**Comment by mattwelborn** [comment]
Attached. Looks like a ZMQ error showed up on this run just now.
[runinfo.tar.gz](https://github.com/Parsl/parsl/files/3684173/runinfo.tar.gz)

**Comment by yadudoc** [comment]
@mattwelborn, I've pushed out a bit more code, could you try running again with those changes. I apologize for how annoying this must be to test. 

I'd also recommend setting `SlurmProvider(cmd_timeout=120)` to give the provider additional time to confirm job submission via slurm. Looks like the default 30s is not sufficient.

**Comment by mattwelborn** [comment]
@yadudoc I'm still seeing the issue with 1 running block, but 0 connected workers. Please find attached the runinfo tarball. So far, there has not been a ZMQ error. I will keep it running and update you if anything changes.

[runinfo.tar.gz](https://github.com/Parsl/parsl/files/3686536/runinfo.tar.gz)

**Comment by mattwelborn** [comment]
Hi @yadudoc. Thanks for your continued efforts on this! Should I give it another try?

**Comment by mattwelborn** [comment]
@yadudoc I've run this with my minimal test case for #1146 and everything worked correctly. I will run overnight on a large workload and see what happens.

**Comment by mattwelborn** [comment]
I've now run for about 14 hours continuously on two clusters without issue.

**Comment by mattwelborn** [comment]
@yadudoc @lnaden This is probably on our end, but has there been a change to the way that tasks/workers/blocks work since release 0.8.0? Running in 0.8.0, I was getting 24 whole-node jobs on slurm with 384 simultaneous tasks. Now, I get 3 whole-node jobs on slurm, still with 384 simultaneous tasks. Again, I think the issue is in how we wrap Parsl, but was wondering if you were aware of any changes that could have caused this.

**Comment by benclifford** [comment]
@mattwelborn there have been some changes - ccing in @annawoodard as she's been driving that

**Comment by Lnaden** [comment]
Its likely related to #1217, so it's something we would need to sit down and adjust on our end likely.

**Comment by yadudoc** [comment]
Scaling granularity is definitely not touched by changes in this PR, but as @benclifford pointed out, changes from Anna's PR touches that functionality and would be useful to test out. So, we should track any concerns there separately.

Once @mattwelborn or @Lnaden confirms, I'll merge this into master.

**Comment by mattwelborn** [comment]
@yadudoc I confirm that this PR has resolved issue #1146 to the best we are able to test it. We'll follow up elsewhere about the blocks/workers/tasks change.

**Comment by yadudoc** [comment]
Perfect, let's get this merged then! 
@mattwelborn thanks very much for your help in testing these out.

**Comment by Lnaden** [comment]
Thank you for drilling into this issue! Its a huge help!

**Pull request merged**